python/src/energuide/extracted_datatypes.py

```diff
@@ -239,66 +239,54 @@ def to_dict(self) -> typing.Dict[str, typing.Optional[float]]:
+        'fill_type_french',
+        'spacer_type_english',
+        'spacer_type_french',
+        'type_english',
+        'type_french',
+        'frame_material_english',
+        'frame_material_french',
+    ]
 
     @classmethod
     def from_data(cls,
-                  window: typing.Dict[str, typing.Any],
+                  window: element.Element,
                   window_code: typing.Dict[str, WindowCode]) -> 'Window':
-        code_id = window.get('constructionTypeCode')
-        code = window_code.get(code_id)
-
-        glazing_types_english = code.glazing_types_english if code is not None else None
-        glazing_types_french = code.glazing_types_french if code is not None else None
-        coatings_tints_english = code.coatings_tints_english if code is not None else None
-        coatings_tints_french = code.coatings_tints_french if code is not None else None
-        fill_type_english = code.fill_type_english if code is not None else None
-        fill_type_french = code.fill_type_french if code is not None else None
-        spacer_type_english = code.spacer_type_english if code is not None else None
-        spacer_type_french = code.spacer_type_french if code is not None else None
-        type_english = code.type_english if code is not None else None
-        type_french = code.type_french if code is not None else None
-        frame_material_english = code.frame_material_english if code is not None else None
-        frame_material_french = code.frame_material_french if code is not None else None
-
-        return Window(
-            label=window['label'],
-            glazing_types_english=glazing_types_english,
-            glazing_types_french=glazing_types_french,
-            coatings_tints_english=coatings_tints_english,
-            coatings_tints_french=coatings_tints_french,
-            fill_type_english=fill_type_english,
-            fill_type_french=fill_type_french,
-            spacer_type_english=spacer_type_english,
-            spacer_type_french=spacer_type_french,
-            type_english=type_english,
-            type_french=type_french,
-            frame_material_english=frame_material_english,
-            frame_material_french=frame_material_french,
-            rsi=window['rsi'],
-            width=(window['width'] * 0.001) if (window['width'] is not None) else None,
-            height=(window['height']* 0.001) if (window['height'] is not None) else None,
-        )
+        code_id = window.xpath('Construction/Type/@idref')
+        code = window_code[code_id[0]] if code_id else None
+
+        code_data = {field: getattr(code, field) if code else None for field in cls._CODE_FIELDS}
+        code_data['label'] = window.findtext('Label')
+        code_data['rsi'] = float(window.xpath('Construction/Type/@rValue')[0])
+        code_data['width'] = float(window.xpath('Measurements/@width')[0]) * 0.001
+        code_data['height'] = float(window.xpath('Measurements/@height')[0]) * 0.001
```

Comment:
I know it was in the original code, but looking at it now those `0.001` seem like magic numbers with no context why the multiplication is being done, should we make those a constant, something like `_MILLIMETRES_MULTIPLIER`?

Reply:
Good call